### PR TITLE
Refuse AppendEntries from other branches

### PR DESCRIPTION
**Pull request by eddyashton**
Fix for #3142. Opening as a draft because it's late on a Friday, and I don't remember exactly what I hacked in to get here.

The issue, in short, is that a malicious host could construct a plausible AppendEntries body using entries from a dead branch (which should have been truncated, and will never be committed), which do not match the range that the enclave intended to describe in its brief AppendEntries header, but which will be accepted (at least partially) by another node, resulting in them reaching a deadend state (holding a suffix they believe is committed, which doesn't match the rest of the service's).

The fix has 2 parts:
- Assume that we know the expected `TxID` for every entry in an AppendEntries (aka - the value the sender intended to send us), and check that matches the entry we received (determined by the view and seqno in the GcmHeader IV). If they don't match, don't apply this entry and return a failure (including an AppendEntries NACK) straight away, without considering more of this message.
- To know the intended `TxID` without modifying the AppendEntries header format to include a dynamic view history, we limit each AppendEntries to contain entries from a single view. So as described in the comment, `term_of_idx` actually describes the view of all transactions in this message (`prev_term` may be different, in which case the first transaction in this message is the first transaction in the new `term_of_idx`).

I'm not sure how the latter plays with `lts_compatibility`, so this may need revisiting. Changing the format is likely more intrusive, but there's an issue in the current approach where the new receiving node is making assumptions from the header that a sender running old code didn't include. I _think_ these will just result in a mismatch ("I thought you told me all of these entries were in term 5, but this includes `2.10`?") that can be successfully resolved, since the sender will eventually send partial AEs which can be accepted? But need to confirm.

The other half of this PR is a unit test which confirms the behaviour, and the fix. This required a lot more plumbing than anticipated! The problem is that the serialisation and parsing of the ledger entries in our unit tests doesn't include the GcmHeader containing a TxID, so doesn't do the same checks that we expect to result in a recognised mismatch in the real nodes. We reproduce this by changing some APIs slightly, to explicitly pass a TxID from Raft (where it is known), for the sake of the tests, though it's already implicitly included in a real service. This is rubbish! Additionally, the abstractions around our ledger mocks and async consensus execution are getting extremely unwieldy, so a lot of this PR is just blind plumbing.

### CODE REVIEW

File: src/consensus/ledger_enclave.h [hunk 1]
```diff
@@ -53,7 +62,9 @@ namespace consensus
       const uint8_t* data,
       size_t size,
       bool globally_committable,
-      bool force_chunk = false)
+      bool force_chunk,
+      kv::Term term,
+      kv::Version index)
```
Comment (jumaffre) [inline src/consensus/ledger_enclave.h:67]: Are these expected to be passed to the host?

Comment (eddyashton) [reply src/consensus/ledger_enclave.h:67]: No. They could be, but I don't think we have a purpose for them, and they're not needed at the moment. The argument is that Raft _knows_ the term and index it is appending; logically it is calling `ledger.append(term, index, data)`, so it should pass those fields in case they're important to the receiver. For the stub ledger used in the unit tests, this is important. For this real implementation it's not (currently). But maybe at some point in the future, it would make sense to do the GcmHeader construction here (still within enclave, but logically part of the ledger representation rather than the 'raw' body)?

Comment (jumaffre) [reply src/consensus/ledger_enclave.h:67]: I didn't see that they were used in the unit test logic, makes sense!

File: src/consensus/aft/raft.h [hunk 2]
```diff
@@ -166,7 +166,7 @@ namespace aft
 
     size_t entry_size_not_limited = 0;
     size_t entry_count = 0;
-    Index entries_batch_size = 1;
+    Index entries_batch_size = 20;
```
Comment (eddyashton) [inline src/consensus/aft/raft.h:169]: This is another drive-by mitigation for #1451. Start with larger batches than 1, so that we have fewer AEs and faster catchup in normal operation.

Comment (achamayou) [reply src/consensus/aft/raft.h:169]: Possibly good to split in its own PR if we thought we wanted to measure the perf impact independently.

File: src/consensus/aft/raft.h [hunk 3]
```diff
@@ -1444,20 +1450,39 @@ namespace aft
...
-        send_append_entries_range(to, start_idx, i);
-        start_idx = std::min(i + 1, state->last_idx);
-      }
+      auto calculate_end_index = [this](Index start) {
+        // Cap the end index in 2 ways:
+        // - Must contain no more than entries_batch_size entries
+        // - Must contain entries from a single term
+        auto max_end = state->last_idx;
+        const auto term_of_ae = state->view_history.view_at(start);
+        const auto end_of_term = state->view_history.end_of_view(term_of_ae);
```
Comment (achamayou) [inline src/consensus/aft/raft.h:1467]: I would suggest index_at_end_of_term instead.

File: src/consensus/aft/raft.h [hunk 4]
```diff
@@ -1444,20 +1450,39 @@ namespace aft
...
-      for (Index i = end_idx; i < state->last_idx; i += entries_batch_size)
-      {
-        send_append_entries_range(to, start_idx, i);
-        start_idx = std::min(i + 1, state->last_idx);
-      }
+      auto calculate_end_index = [this](Index start) {
+        // Cap the end index in 2 ways:
+        // - Must contain no more than entries_batch_size entries
+        // - Must contain entries from a single term
+        auto max_end = state->last_idx;
```
Comment (achamayou) [inline src/consensus/aft/raft.h:1465]: max_index?

**Review by achamayou — APPROVED** [review]
